### PR TITLE
Add estimator for object tracker GOTURN in PyTorch

### PR DESCRIPTION
**Pull request by beat-buesser**
# Description

This pull request add the first estimator for object tracking models supporting GOTURN in PyTorch.

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: .github/actions/goturn/action.yml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+name: 'Test DeepSpeech v3'
```
Comment (minhitbk) [inline .github/actions/goturn/action.yml:1]: Please update this line

File: .github/actions/goturn/action.yml [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+description: 'Run tests for DeepSpeech v3'
```
Comment (minhitbk) [inline .github/actions/goturn/action.yml:2]: Please update this line

File: art/attacks/evasion/fast_gradient.py [hunk 3]
```diff
@@ -472,7 +472,11 @@ def _apply_perturbation(
         batch = batch + perturbation_step
         if self.estimator.clip_values is not None:
             clip_min, clip_max = self.estimator.clip_values
-            batch = np.clip(batch, clip_min, clip_max)
+            if batch.dtype == np.object:
```
Comment (minhitbk) [inline art/attacks/evasion/fast_gradient.py:475]: Shouldn't we use numpy?

File: art/attacks/evasion/fast_gradient.py [hunk 4]
```diff
@@ -472,7 +472,11 @@ def _apply_perturbation(
...
+                for i_obj in range(batch.shape[0]):
+                    batch[i_obj] = np.clip(batch[i_obj], clip_min, clip_max)
+            else:
+                batch = np.clip(batch, clip_min, clip_max)
```
Comment (minhitbk) [inline art/attacks/evasion/fast_gradient.py:479]: Shouldn't we use numpy?

File: art/attacks/evasion/fast_gradient.py [hunk 5]
```diff
@@ -472,7 +472,11 @@ def _apply_perturbation(
         batch = batch + perturbation_step
         if self.estimator.clip_values is not None:
             clip_min, clip_max = self.estimator.clip_values
-            batch = np.clip(batch, clip_min, clip_max)
+            if batch.dtype == np.object:
+                for i_obj in range(batch.shape[0]):
+                    batch[i_obj] = np.clip(batch[i_obj], clip_min, clip_max)
```
Comment (minhitbk) [inline art/attacks/evasion/fast_gradient.py:477]: Shouldn't we use numpy?

File: art/estimators/object_tracking/pytorch_goturn.py [hunk 6]
```diff
@@ -0,0 +1,626 @@
...
+        :param x: Samples of shape (nb_samples, nb_frames, height, width, nb_channels).
+        :param y: Target values of format `List[Dict[str, np.ndarray]]`, one dictionary for each input image. The keys
+                  of the dictionary are:
+
+                  - boxes [N_FRAMES, 4]: the boxes in [x1, y1, x2, y2] format, with 0 <= x1 < x2 <= W and
+                                         0 <= y1 < y2 <= H.
+                  - labels [N_FRAMES]: the labels for each image, default 0.
+                  - scores [N_FRAMES]: the scores or each prediction, default 1.
+        :param reduction: Specifies the reduction to apply to the output: 'none' | 'mean' | 'sum'.
+                   'none': no reduction will be applied
```
Comment (minhitbk) [inline art/estimators/object_tracking/pytorch_goturn.py:183]:
```suggestion
                   'none': no reduction will be applied.
```

File: art/estimators/object_tracking/pytorch_goturn.py [hunk 7]
```diff
@@ -0,0 +1,626 @@
...
+                   'sum': the output will be summed.
```
Comment (minhitbk) [inline art/estimators/object_tracking/pytorch_goturn.py:184]: Should we add one line for the 'mean'

File: art/estimators/object_tracking/pytorch_goturn.py [hunk 8]
```diff
@@ -0,0 +1,626 @@
...
+                y_tensor = y
+
+            # transform = torchvision.transforms.Compose([torchvision.transforms.ToTensor()])
+            image_tensor_list_grad = list()
+            y_preprocessed = list()
+            inputs_t = list()
+
+            for i in range(x.shape[0]):
+                if self.clip_values is not None:
+                    # x_grad = transform(x[i] / self.clip_values[1]).to(self._device)
```
Comment (minhitbk) [inline art/estimators/object_tracking/pytorch_goturn.py:216]: Why don't we remove this line?

File: art/estimators/object_tracking/pytorch_goturn.py [hunk 9]
```diff
@@ -0,0 +1,626 @@
...
+                    x_grad = torch.from_numpy(x[i]).to(self._device).float()
+                else:
+                    # x_grad = transform(x[i]).to(self._device)
```
Comment (minhitbk) [inline art/estimators/object_tracking/pytorch_goturn.py:219]: Why don't we remove this line?

File: art/estimators/object_tracking/pytorch_goturn.py [hunk 10]
```diff
@@ -0,0 +1,626 @@
...
+    ) -> np.ndarray:
+        """
+        Compute the gradient of the loss function w.r.t. `x`.
+
+        :param x: Samples of shape (nb_samples, height, width, nb_channels).
+        :param y: Target values of format `List[Dict[Tensor]]`, one for each input image. The
+                  fields of the Dict are as follows:
+
+                  - boxes (FloatTensor[N, 4]): the predicted boxes in [x1, y1, x2, y2] format, with values \
+                    between 0 and H and 0 and W
```
Comment (minhitbk) [inline art/estimators/object_tracking/pytorch_goturn.py:272]:
```suggestion
                    between 0 and H and 0 and W.
```

File: art/estimators/object_tracking/pytorch_goturn.py [hunk 11]
```diff
@@ -0,0 +1,626 @@
...
+                  - labels (Int64Tensor[N]): the predicted labels for each image
```
Comment (minhitbk) [inline art/estimators/object_tracking/pytorch_goturn.py:273]:
```suggestion
                  - labels (Int64Tensor[N]): the predicted labels for each image.
```

File: .github/actions/goturn/action.yml (hunk 1 above)
Comment (beat-buesser) [reply .github/actions/goturn/action.yml:1]: Ok.

File: .github/actions/goturn/action.yml (hunk 2 above)
Comment (beat-buesser) [reply .github/actions/goturn/action.yml:2]: Ok.

File: art/estimators/object_tracking/pytorch_goturn.py (hunk 7 above)
Comment (beat-buesser) [reply art/estimators/object_tracking/pytorch_goturn.py:184]: Ok.

File: art/estimators/object_tracking/pytorch_goturn.py (hunk 8 above)
Comment (beat-buesser) [reply art/estimators/object_tracking/pytorch_goturn.py:216]: Yes.

File: art/estimators/object_tracking/pytorch_goturn.py (hunk 9 above)
Comment (beat-buesser) [reply art/estimators/object_tracking/pytorch_goturn.py:219]: Yes.

File: art/attacks/evasion/fast_gradient.py (hunk 3 above)
Comment (beat-buesser) [reply art/attacks/evasion/fast_gradient.py:475]: This change is to handle object arrays of numpy arrays.

File: art/estimators/object_tracking/pytorch_goturn.py [hunk 12]
```diff
@@ -0,0 +1,626 @@
...
+
+        if x.dtype != object:
+            grads = np.array([i for i in grads], dtype=x.dtype)  # pylint: disable=R1721
+            assert grads.shape == x.shape and grads.dtype == x.dtype
+
+        return grads
+
+    def _preprocess(self, img):
+        """
+        preprocess image before forward pass, this is the same
```
Comment (minhitbk) [inline art/estimators/object_tracking/pytorch_goturn.py:319]:
```suggestion
        Preprocess image before forward pass, this is the same
```

File: art/estimators/object_tracking/pytorch_goturn.py [hunk 13]
```diff
@@ -0,0 +1,626 @@
...
+        preprocessing used during training, please refer to collate function
+        in train.py for reference
+        @image: input image
```
Comment (minhitbk) [inline art/estimators/object_tracking/pytorch_goturn.py:322]: Can you reformat lines 322-325 consistently with ART style?

File: art/estimators/object_tracking/pytorch_goturn.py [hunk 14]
```diff
@@ -0,0 +1,626 @@
...
+        img = img.permute(2, 0, 1)
+        img = img * std + mean
+        img = torch.unsqueeze(img, dim=0)
+        img = interpolate(img, size=(227, 227))
+        img = torch.squeeze(img)
+        img = (img - mean) / std
+        return img
+
+    def _track(self, curr_frame, prev_frame, rect):
+        """track current frame
```
Comment (minhitbk) [inline art/estimators/object_tracking/pytorch_goturn.py:340]: Can you reformat consistently with ART style?, also for other following functions.

File: art/estimators/object_tracking/pytorch_goturn.py [hunk 15]
```diff
@@ -0,0 +1,626 @@
...
+        if not self.all_framework_preprocessing:
+            grads = self._apply_preprocessing_gradient(x, grads)
+
+        if x.dtype != object:
+            grads = np.array([i for i in grads], dtype=x.dtype)  # pylint: disable=R1721
+            assert grads.shape == x.shape and grads.dtype == x.dtype
+
+        return grads
+
+    def _preprocess(self, img):
```
Comment (minhitbk) [inline art/estimators/object_tracking/pytorch_goturn.py:317]: Can you write with python3 style?

File: art/estimators/object_tracking/pytorch_goturn.py [hunk 16]
```diff
@@ -0,0 +1,626 @@
...
+        std = torch.from_numpy(std_np).reshape((3, 1, 1))
+        img = img.permute(2, 0, 1)
+        img = img * std + mean
+        img = torch.unsqueeze(img, dim=0)
+        img = interpolate(img, size=(227, 227))
+        img = torch.squeeze(img)
+        img = (img - mean) / std
+        return img
+
+    def _track(self, curr_frame, prev_frame, rect):
```
Comment (minhitbk) [inline art/estimators/object_tracking/pytorch_goturn.py:339]: Can you write with python3 style? Please change for all following functions.

File: art/estimators/object_tracking/pytorch_goturn.py [hunk 17]
```diff
@@ -0,0 +1,626 @@
...
+            """
+            return (bbox_tight[0] + bbox_tight[2]) / 2.0
+
+        def get_center_y_f(bbox_tight):
+            """
+            y-coordinate of the bounding box center
+            """
+            return (bbox_tight[1] + bbox_tight[3]) / 2.0
+
+        def compute_crop_pad_image_location(bbox_tight, image):
```
Comment (minhitbk) [inline art/estimators/object_tracking/pytorch_goturn.py:381]: Is this static function?, please also change with python3 style

File: art/estimators/object_tracking/pytorch_goturn.py [hunk 18]
```diff
@@ -0,0 +1,626 @@
...
+        def compute_output_height_f(bbox_tight):
+            """
+            height of search/target region
+            """
+            bbox_height = bbox_tight[3] - bbox_tight[1]
+            output_height = k_context_factor * bbox_height
+
+            return max(1.0, output_height)
+
+        def compute_output_width_f(bbox_tight):
```
Comment (minhitbk) [inline art/estimators/object_tracking/pytorch_goturn.py:360]: Is this static function?, please also change with python3 style

File: art/estimators/object_tracking/pytorch_goturn.py [hunk 19]
```diff
@@ -0,0 +1,626 @@
...
+            """
+            width of search/target region
+            """
+            bbox_width = bbox_tight[2] - bbox_tight[0]
+            output_width = k_context_factor * bbox_width
+
+            return max(1.0, output_width)
+
+        def get_center_x_f(bbox_tight):
```
Comment (minhitbk) [inline art/estimators/object_tracking/pytorch_goturn.py:369]: Is this static function?, please also change with python3 style

File: art/estimators/object_tracking/pytorch_goturn.py [hunk 20]
```diff
@@ -0,0 +1,626 @@
...
+        pred_bb[1] = max(0.0, pred_bb[1] + search_location[1] - edge_spacing_y)
+        pred_bb[2] = min(raw_image.shape[1], pred_bb[2] + search_location[0] - edge_spacing_x)
+        pred_bb[3] = min(raw_image.shape[0], pred_bb[3] + search_location[1] - edge_spacing_y)
+
+        # brings gradients to zero
+        # pred_bb = torch.round(pred_bb)
+
+        return pred_bb
+
+    def track(self, x, y_init):
```
Comment (minhitbk) [inline art/estimators/object_tracking/pytorch_goturn.py:528]: please also change with python3 style

File: art/estimators/object_tracking/pytorch_goturn.py [hunk 21]
```diff
@@ -0,0 +1,626 @@
...
+
+        # move the bounding box to target/search region coordinates
+        raw_image = curr_frame
+        pred_bb[0] = max(0.0, pred_bb[0] + search_location[0] - edge_spacing_x)
+        pred_bb[1] = max(0.0, pred_bb[1] + search_location[1] - edge_spacing_y)
+        pred_bb[2] = min(raw_image.shape[1], pred_bb[2] + search_location[0] - edge_spacing_x)
+        pred_bb[3] = min(raw_image.shape[0], pred_bb[3] + search_location[1] - edge_spacing_y)
+
+        # brings gradients to zero
+        # pred_bb = torch.round(pred_bb)
```
Comment (minhitbk) [inline art/estimators/object_tracking/pytorch_goturn.py:524]: Please remove all comment lines if they are not needed.

File: art/estimators/object_tracking/pytorch_goturn.py [hunk 22]
```diff
@@ -0,0 +1,626 @@
...
+        """
+        import torch  # lgtm [py/repeated-import]
+
+        prev_bbox = rect
+
+        k_context_factor = 2
+
+        def compute_output_height_f(bbox_tight):
+            """
+            height of search/target region
```
Comment (minhitbk) [inline art/estimators/object_tracking/pytorch_goturn.py:353]:
```suggestion
            Height of search/target region.
```

File: art/estimators/object_tracking/pytorch_goturn.py [hunk 23]
```diff
@@ -0,0 +1,626 @@
...
+            """
+            bbox_height = bbox_tight[3] - bbox_tight[1]
+            output_height = k_context_factor * bbox_height
+
+            return max(1.0, output_height)
+
+        def compute_output_width_f(bbox_tight):
+            """
+            width of search/target region
```
Comment (minhitbk) [inline art/estimators/object_tracking/pytorch_goturn.py:362]:
```suggestion
            Width of search/target region.
```

**Review by minhitbk — CHANGES_REQUESTED** [review]
@beat-buesser, I only have some minor comments. In addition, I suggest you reformat code in the _track function, especially those subfunctions, so that they obey ART style.

File: art/estimators/object_tracking/pytorch_goturn.py (hunk 13 above)
Comment (beat-buesser) [reply art/estimators/object_tracking/pytorch_goturn.py:322]: Ok.

File: art/estimators/object_tracking/pytorch_goturn.py (hunk 14 above)
Comment (beat-buesser) [reply art/estimators/object_tracking/pytorch_goturn.py:340]: Ok.

File: art/estimators/object_tracking/pytorch_goturn.py (hunk 15 above)
Comment (beat-buesser) [reply art/estimators/object_tracking/pytorch_goturn.py:317]: Ok.

File: art/estimators/object_tracking/pytorch_goturn.py (hunk 16 above)
Comment (beat-buesser) [reply art/estimators/object_tracking/pytorch_goturn.py:339]: Ok.

File: art/estimators/object_tracking/pytorch_goturn.py (hunk 20 above)
Comment (beat-buesser) [reply art/estimators/object_tracking/pytorch_goturn.py:528]: Ok.

File: art/estimators/object_tracking/pytorch_goturn.py (hunk 21 above)
Comment (beat-buesser) [reply art/estimators/object_tracking/pytorch_goturn.py:524]: I agree and will remove them in the upcoming patch releases. I would like to keep them for now for documentation and future debugging.

**Comment by beat-buesser** [comment]
@minhitbk Thank you very much!

**Pull request merged**